src/libraries/System.Private.CoreLib/src/System/CoreLib.cs

```diff
@@ -2,11 +2,53 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
 namespace System
 {
-    // This class is used to define the name of the base class library
+    /// <summary>
+    /// Defines capabilities that cut across the whole CoreLib library.
+    /// </summary>
     internal static class CoreLib
     {
+        /// <summary>
+        /// Defines the name of the base class library
+        /// </summary>
         public const string Name = "System.Private.CoreLib";
+
+        /// <summary>
+        /// InitializeILLinkTrimmableCode is a method that is never called, but exists
+        /// in order to define feature switches for the ILLinker to analyze.
+        /// An application can set these feature switches to tell the linker to trim code
+        /// that is not needed by the application.
+        /// </summary>
+        private static void InitializeILLinkTrimmableCode()
+        {
+            if (!DebugMode.None)
+            {
+                InitializeDebugSupport();
+            }
+        }
+
+        [PreserveDependency("get_ParentForDebugger", "System.Threading.Tasks.Task")]
+        [PreserveDependency("get_StateFlagsForDebugger", "System.Threading.Tasks.Task")]
+        [PreserveDependency("GetDelegateContinuationsForDebugger", "System.Threading.Tasks.Task")]
+        [PreserveDependency("SetNotificationForWaitCompletion", "System.Threading.Tasks.Task")]
+        [PreserveDependency("GetActiveTaskFromId", "System.Threading.Tasks.Task")] // used by VS Tasks Window
+        [PreserveDependency("GetQueuedWorkItemsForDebugger", "System.Threading.ThreadPool")]
+        [PreserveDependency("GetGloballyQueuedWorkItemsForDebugger", "System.Threading.ThreadPool")]
+        [PreserveDependency("GetLocallyQueuedWorkItemsForDebugger", "System.Threading.ThreadPool")]
+        [PreserveDependency("TryGetStateMachineForDebugger", "System.Runtime.CompilerServices.AsyncMethodBuilderCore")]
+        [PreserveDependency("get_ObjectIdForDebugger", "System.Runtime.CompilerServices.AsyncIteratorMethodBuilder")]
+        [PreserveDependency("get_ObjectIdForDebugger", "System.Runtime.CompilerServices.AsyncVoidMethodBuilder")]
+        [PreserveDependency("get_ObjectIdForDebugger", "System.Runtime.CompilerServices.AsyncTaskMethodBuilder")]
+        [PreserveDependency("SetNotificationForWaitCompletion", "System.Runtime.CompilerServices.AsyncTaskMethodBuilder")]
+        private static void InitializeDebugSupport()
```

Comment:
I'm not sure this is setting the right precedence I think we should move `PreserveDependency` closer to the type as it does not make sense to keep the member if the type is not needed (yeah, we'll need conditional support but let's add it)

Reply:
> yeah, we'll need conditional support but let's add it

I logged https://github.com/mono/linker/issues/1231 to get this conditional support added.

> I'm not sure this is setting the right precedence I think we should move PreserveDependency closer to the type as it does not make sense to keep the member if the type is not needed

The problem is there is no good place to put the attribute. `DynamicDependency` only allows:

https://github.com/dotnet/runtime/blob/ec6bec455514070cc69e4120974796338bcb7cb7/src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/DynamicDependencyAttribute.cs#L14-L17

Putting it on one of the constructors of `Task` type doesn't make sense, because that constructor may get trimmed. Same for any other member of the `Task` type.

Where do you think they should be placed?

BTW - @vitek-karas and I were thinking on using this same pattern for COM support. We would just have another feature switch for COM, and putting `if (COMSupported) InitializeCOMSupport();`, where `InitializeCOMSupport()` will have the `DynamicDependency` attributes to all the types that are necessary for COM.

Reply:
I think in the Task case you could add it to constructor but for the structs it's tricky. We could also extend DynamicDependencyAttribute targets

Reply:
I think that in general the pattern which is introduced here is a good one - for things which are "global" - so things which are needed by the runtime itself and linker won't be able to tell if they're in-use or not. I think COM is a good example, but I might be wrong.

For things which are only needed if some other managed class/method is used we should try to move the dependencies there - to improve trimability. We do care about size even for apps which have enabled debugger support for example.

I don't have a problem with both:
* Adding feature support to `DynamicDependencyAttribute`
* Allowing `DynamicDependencyAttribute` on types (the meaning would be that if we decide to keep the type in any way or form, those dependencies would kick in)

To make progress I think that we should take the changes as Eric wrote them here, it's an improvement as it makes the debugger stuff conditional on a feature switch. And once we have the necessary bits in the linker, we can improve on it by moving more of the local dependencies closer to where they belong.

Reply:
I've opened https://github.com/dotnet/runtime/issues/37352 to allow `[DynamicDependency]` attributes on a type.

We will need the ILLinker to support both #37352 and https://github.com/mono/linker/issues/1231 before we can take this approach.

Reply:
> To make progress I think that we should take the changes as Eric wrote them here

It might be better to split this into 4 PRs
1. support for multiple substitutions per assembly
2. unification of XML descriptors for SPC
3. introduction of Debugger.IsSupported or similar and using it in libraries (for example with https://github.com/dotnet/runtime/blob/8deadd2de5ee1c220c3cf851c474c3717a944167/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs#L5816)
4. Conversion of XML decriptors for the debugger to attributes

Reply:
PR for `1. support for multiple substitutions per assembly` is here: https://github.com/dotnet/runtime/pull/37443